src/ws.nim

```diff
@@ -0,0 +1,415 @@
+      i += maxSize
+  except IOError, OSError, ValueError:
+    # Wrap all exceptions in a WebSocketError so its easy to catch
+    raise newException(WebSocketError, "Failed to send data: " &
+        getCurrentExceptionMsg())
+
+proc sendStr*(ws: WebSocket, data: string, opcode = Opcode.Text): Future[
+                void] {.async.} =
+  await send(ws, toBytes(data), opcode)
+
+proc close*(ws: WebSocket) =
+  ## Close the Socket, sends close packet.
+  ws.readyState = Closed
+  proc close() {.async.} =
+    await ws.send(@[], Close)
+    ws.tcpSocket.close()
+  asyncCheck close()
+
+proc receiveFrame(ws: WebSocket): Future[Frame] {.async.} =
+  ## Gets a frame from the WebSocket.
+  ## See https://tools.ietf.org/html/rfc6455#section-5.2
+
+  # Grab the header.
+  var header = newSeq[byte](2)
+  try:
+    await ws.tcpSocket.readExactly(addr header[0], 2)
+  except TransportUseClosedError:
+    ws.readyState = Closed
+    raise newException(WebSocketError, "Socket closed")
+  except CatchableError:
+    ws.readyState = Closed
+    raise newException(WebSocketError, "Failed to read websocket header")
+
+  if header.len != 2:
+    ws.readyState = Closed
+    raise newException(WebSocketError, "Invalid websocket header length")
+
+  let b0 = header[0].uint8
+  let b1 = header[1].uint8
+
+  var frame: Frame
+  # Read the flags and fin from the header.
+
+  var hf = cast[HeaderFlags](b0 shr 4)
+  frame.fin = fin in hf
+  frame.rsv1 = rsv1 in hf
+  frame.rsv2 = rsv2 in hf
+  frame.rsv3 = rsv3 in hf
+
+  var opcode = b0 and 0x0f
+  if opcode notin WSOpCode:
+    raise newException(WebSocketError, "Unexpected  websocket opcode")
+  frame.opcode = (opcode).Opcode
+
+  # If any of the rsv are set close the socket.
+  if frame.rsv1 or frame.rsv2 or frame.rsv3:
+    ws.readyState = Closed
+    raise newException(WebSocketError, "WebSocket rsv mismatch")
+
+  # Payload length can be 7 bits, 7+16 bits, or 7+64 bits.
+  var finalLen: uint64 = 0
+
+  let headerLen = uint(b1 and 0x7f)
+  if headerLen == 0x7e:
+    # Length must be 7+16 bits.
+    var length = newSeq[byte](2)
+    await ws.tcpSocket.readExactly(addr length[0], 2)
+    finalLen = cast[ptr uint16](length[0].addr)[].toBE
+  elif headerLen == 0x7f:
+    # Length must be 7+64 bits.
+    var length = newSeq[byte](8)
+    await ws.tcpSocket.readExactly(addr length[0], 8)
+    finalLen = cast[ptr uint64](length[0].addr)[].toBE
+  else:
+    # Length must be 7 bits.
+    finalLen = headerLen
+
+  # Do we need to apply mask?
+  frame.mask = (b1 and 0x80) == 0x80
+
+  if ws.masked == frame.mask:
+    # Server sends unmasked but accepts only masked.
+    # Client sends masked but accepts only unmasked.
+    raise newException(WebSocketError, "Socket mask mismatch")
+
+  var maskKey = newSeq[byte](4)
+  if frame.mask:
+    # Read the mask.
+    await ws.tcpSocket.readExactly(addr maskKey[0], 4)
+
+  # Read the data.
+  var data = newSeq[byte](finalLen)
```

Comment:
**SECURITY WARNING** I dont see any checks on `finalLen` size before allocating chunk of memory... So looks like its possible to allocate `high(uint64)` bytes of memory with frame which will be just 9 bytes size.

Reply:
Added `MsgReader` that fills the buffer provided by the user of the library.
```
proc readMessage*(msgReader: MsgReader,data: seq[byte]): MsgReader {.async.} =
```